### ISSUE TITLE
`unfulfilled_lint_expectations` emitted upon a correct `#[expect(clippy::enum_glob_use)]`

### ISSUE THREAD

**Issue by its-the-shrimp** [rust-lang/rust-analyzer#17685]
**rust-analyzer version**: `2024-07-22`

**rustc version**: rustc 1.82.0-nightly (8bfcae730 2024-07-23)

**editor or extension**: Neovim

**code snippet to reproduce**: 
```rust
#![warn(clippy::pedantic)]

#[expect(clippy::enum_glob_use)]
use std::cmp::Ordering::*;

fn main() {
    print!("{Equal:?}");
}
```
This code is just fine with `cargo check` or `cargo clippy`, but rust-analyzer emits an `unfulfilled_lint_expectations` warning, even though the `clippy::enum_glob_use` lint is triggered if the `expect` attribute is commented out

**Comment by kpreid** [comment]
I'm seeing the same thing with other lints, e.g.

```rust
//! crate docs

#![warn(missing_docs)]

#[expect(missing_docs)] // warning: this lint expectation is unfulfilled
pub fn foo() {}
```

rust-analyzer version: 0.3.2086-standalone (7106cd3be 2024-08-25) [/Users/kpreid/.vscode/extensions/rust-lang.rust-analyzer-0.3.2086-darwin-x64/server/rust-analyzer]
rustc version: 1.82.0-nightly (c6db1ca3c 2024-08-25)

Presumably this has something to do with how RA invokes the check command, but I couldn't figure out how to extract the command line to test it separately. It happens whether `rust-analyzer.check.command` is set to `check` or `clippy`.

**Comment by MrCroxx** [comment]
> I'm seeing the same thing with other lints, e.g.
> 
> ```rust
> //! crate docs
> 
> #![warn(missing_docs)]
> 
> #[expect(missing_docs)] // warning: this lint expectation is unfulfilled
> pub fn foo() {}
> ```
> 
> rust-analyzer version: 0.3.2086-standalone ([7106cd3](https://github.com/rust-lang/rust-analyzer/commit/7106cd3be50b2a43c1d9f2787bf22d4369c2b25b) 2024-08-25) [/Users/kpreid/.vscode/extensions/rust-lang.rust-analyzer-0.3.2086-darwin-x64/server/rust-analyzer] rustc version: 1.82.0-nightly (c6db1ca3c 2024-08-25)
> 
> Presumably this has something to do with how RA invokes the check command, but I couldn't figure out how to extract the command line to test it separately. It happens whether `rust-analyzer.check.command` is set to `check` or `clippy`.

+1 with rust stable-1.81.0

**Comment by Victor-N-Suadicani** [comment]
Just updated to stable 1.81 and got this.

I can reproduce it directly with cargo using `cargo clippy --tests`, which also reports "this lint expectation is unfulfilled". But it doesn't happen with `cargo check --tests`. :thinking:

**Comment by lnicola** [comment]
This happens because we pass `--all-targets` by default. But I don't really understand the reason why, we should probably file an issue against Clippy.

**Comment by Urgau** [comment]
`--all-targets` enables `--tests` which enables the test harness generation (`rustc --test`) which some lint are gated against, which means that they don't emit the lint which triggers the `unfulfilled_lint_expectations` lint since it wasn't emitted

**Comment by ChayimFriedman2** [comment]
As said this is enabled by `--all-targets` as well. [You can disable `--all-targets` if you want](https://rust-analyzer.github.io/book/configuration.html#cargo.allTargets). Other than that, there is not much rust-analyzer can do.
